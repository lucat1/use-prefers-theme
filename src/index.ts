import { useState, useEffect } from 'react'

export type Preference = 'dark' | 'light' | 'none'

// makeMatch executes the media query to check the preference
const makeMatch = () => window.matchMedia('(prefers-color-scheme: dark)')

// pref translates into our Preference type the value of the media query
const pref = (match: MediaQueryList): Preference =>
  match.matches ? 'dark' : 'light'

// usePrefersTheme uses the media query api to check if
// the user prefers a dark or light colorscheme
// changing the system setting will trigger a rerener in react
export const usePrefersTheme = (): Preference => {
  const supported = typeof window !== 'undefined' && 'matchMedia' in window

  // first we check for environments which don't support
  // the required api (server-side, old browsers)
  if (!supported) {
    console.log(window.matchMedia)
    return 'none'
  }

  // at this point we are sure we got the requied api so we
  // can implement the real hook logic

  // firstly we get the value at render time
  // we save the match variable to reuse the same query later on
  // in the event listener logic
  const match = makeMatch()
  const [preference, setPreference] = useState(pref(match))

  useEffect(() => {
    const handler = () => {
      setPreference(pref(match))
    }

    match.addEventListener('change', handler)
    return () => match.removeEventListener('change', handler)
  }, [match])

  return preference
}

export default usePrefersTheme
