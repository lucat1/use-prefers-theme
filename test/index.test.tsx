import * as React from 'react'
import * as ReactDOM from 'react-dom'
import * as ReactTest from 'react-test-renderer'
import { usePrefersTheme } from '../src'

const TestComponent = () => {
  const pref = usePrefersTheme()
  return <div>{pref}</div>
}

describe('it', () => {
  it('renders in a browser', () => {
    const div = document.createElement('div')
    ReactDOM.render(<TestComponent />, div)
    ReactDOM.unmountComponentAtNode(div)

    expect(ReactTest.create(<TestComponent />)).toMatchSnapshot()
  })

  it('renders in a ssr environment', () => {
    ;(window as any) = undefined
    const div = document.createElement('div')
    ReactDOM.render(<TestComponent />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('renders in an old environment', () => {
    delete window.matchMedia
    const div = document.createElement('div')
    ReactDOM.render(<TestComponent />, div)
    ReactDOM.unmountComponentAtNode(div)
  })
})

describe('it renders in browsers and', () => {
  let rendered: ReactTest.ReactTestRenderer
  const mount = jest.fn(),
    unmount = jest.fn()
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: true, // = dark
      media: query,
      onchange: null,
      addEventListener: mount,
      removeEventListener: unmount,
      dispatchEvent: jest.fn()
    }))
  })

  it('returns dark/light value', () => {
    rendered = ReactTest.create(<TestComponent />)
    const json = rendered.toJSON()
    if (json == null) {
      fail('json render is null')
    }

    expect(window.matchMedia).toBeCalled()

    expect(json.children).not.toBeNull()
    expect(json.children).toHaveLength(1)
    const children = json.children as ReactTest.ReactTestRendererNode[]
    expect(children[0]).toMatch('dark')
  })

  it('calls addEventListener', () => {
    expect(mount).toBeCalled()
  })

  it('calls removeEventListener', () => {
    rendered.unmount()
    expect(window.matchMedia('').removeEventListener).toBeCalled()
  })
})
