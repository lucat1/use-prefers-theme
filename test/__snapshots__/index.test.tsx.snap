// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders in a browser 1`] = `
<div>
  dark
</div>
`;
